---
title: "Data Cleaning"
format: html
---

# Record Data Cleaning Methods

In this section I will clean and merge all my record data so that I will be left with one table that I will use to build my models. I will also reconfigure my data so it will be suitable for my analysis. 

**Important: It is at this point I discovered my project was based on time series data. For the rest of the project I needed my data to follow the IID assumption (independent and identically distributed). The time series data I have is not independent, so I pivoted how I would use my data. An observational unit in my case is a specific state coupled with a specific year, like Alabama-2019. The target variable, or variable I am trying to predict, is the percent increase in home value from one year to the next. Below I will explain what a single row of my finished table means more thoroughly. Although all my data is not independent, so you will need to be wary of the validity of some of my conclusions, you can still gain useful information about real estate trends from my analysis and models I construct.**

In the code below, you will see me read in specific columns from each table, change column names to accurately represent what is in the column, merge and concatenate all the tables together as one, drop NA values that result from the merges, convert each cell to percent change from the previous year, and check for potential problems with the data, specifically outliers.

# Record Data Cleaning Code

In [70]:
import requests
import json
import re
import pandas as pd
DP02cols=['NAME','DP02_0001E','DP02_0002E','DP02_0003E','DP02_0007E','DP02_0011E','DP02_0037E','DP02_0060E','DP02_0061E','DP02_0062E','DP02_0063E','DP02_0064E','DP02_0065E']
DP03cols=['NAME','DP03_0026PE','DP03_0005E','DP03_0062E','DP03_0063E','DP03_0093E','DP03_0094E']
DP04cols=['NAME','DP04_0002E','DP04_0001E','DP04_0037E','DP04_0047E','DP04_0134E']
DP05cols=['NAME','DP05_0001E','DP05_0004E','DP05_0018E','DP05_0037E','DP05_0038E','DP05_0039E','DP05_0044E','DP05_0071E']

DP0217=pd.read_csv('data/2017DP02.csv',skiprows=1)
DP0218=pd.read_csv('data/2018DP02.csv',skiprows=1)
DP0219=pd.read_csv('data/2019DP02.csv',skiprows=1)
DP0221=pd.read_csv('data/2021DP02.csv',skiprows=1)
DP0222=pd.read_csv('data/2022DP02.csv',skiprows=1)

DP0317 = pd.read_csv('data/2017DP03.csv', skiprows=1)
DP0318 = pd.read_csv('data/2018DP03.csv', skiprows=1)
DP0319 = pd.read_csv('data/2019DP03.csv', skiprows=1)
DP0321 = pd.read_csv('data/2021DP03.csv', skiprows=1)
DP0322 = pd.read_csv('data/2022DP03.csv', skiprows=1)

DP0417 = pd.read_csv('data/2017DP04.csv', skiprows=1)
DP0418 = pd.read_csv('data/2018DP04.csv', skiprows=1)
DP0419 = pd.read_csv('data/2019DP04.csv', skiprows=1)
DP0421 = pd.read_csv('data/2021DP04.csv', skiprows=1)
DP0422 = pd.read_csv('data/2022DP04.csv', skiprows=1)

DP0517 = pd.read_csv('data/2017DP05.csv', skiprows=1)
DP0518 = pd.read_csv('data/2018DP05.csv', skiprows=1)
DP0519 = pd.read_csv('data/2019DP05.csv', skiprows=1)
DP0521 = pd.read_csv('data/2021DP05.csv', skiprows=1)
DP0522 = pd.read_csv('data/2022DP05.csv', skiprows=1)

DP0217=DP0217[DP02cols]
DP0217['Year']=2017
DP0218=DP0218[DP02cols]
DP0218['Year']=2018
DP0219=DP0219[DP02cols]
DP0219['Year']=2019
DP0221=DP0221[DP02cols]
DP0221['Year']=2021
DP0222=DP0222[DP02cols]
DP0222['Year']=2022

DP0317 = DP0317[DP03cols]
DP0318 = DP0318[DP03cols]
DP0319 = DP0319[DP03cols]
DP0321 = DP0321[DP03cols]
DP0322 = DP0322[DP03cols]

DP0417 = DP0417[DP04cols]
DP0418 = DP0418[DP04cols]
DP0419 = DP0419[DP04cols]
DP0421 = DP0421[DP04cols]
DP0422 = DP0422[DP04cols]

DP0517 = DP0517[DP05cols]
DP0518 = DP0518[DP05cols]
DP0519 = DP0519[DP05cols]
DP0521 = DP0521[DP05cols]
DP0522 = DP0522[DP05cols]

df2017=pd.merge(DP0217,DP0317,on='NAME')
df2017=pd.merge(df2017,DP0417,on='NAME')
df2017=pd.merge(df2017,DP0517,on='NAME')

df2018=pd.merge(DP0218,DP0318,on='NAME')
df2018=pd.merge(df2018,DP0418,on='NAME')
df2018=pd.merge(df2018,DP0518,on='NAME')

df2019=pd.merge(DP0219,DP0319,on='NAME')
df2019=pd.merge(df2019,DP0419,on='NAME')
df2019=pd.merge(df2019,DP0519,on='NAME')


df2021=pd.merge(DP0221,DP0321,on='NAME')
df2021=pd.merge(df2021,DP0421,on='NAME')
df2021=pd.merge(df2021,DP0521,on='NAME')

df2022=pd.merge(DP0222,DP0322,on='NAME')
df2022=pd.merge(df2022,DP0422,on='NAME')
df2022=pd.merge(df2022,DP0522,on='NAME')

In [71]:
df2017.columns

Index(['NAME', 'DP02_0001E', 'DP02_0002E', 'DP02_0003E', 'DP02_0007E',
       'DP02_0011E', 'DP02_0037E', 'DP02_0060E', 'DP02_0061E', 'DP02_0062E',
       'DP02_0063E', 'DP02_0064E', 'DP02_0065E', 'Year', 'DP03_0026PE',
       'DP03_0005E', 'DP03_0062E', 'DP03_0063E', 'DP03_0093E', 'DP03_0094E',
       'DP04_0002E', 'DP04_0001E', 'DP04_0037E', 'DP04_0047E', 'DP04_0134E',
       'DP05_0001E', 'DP05_0004E', 'DP05_0018E', 'DP05_0037E', 'DP05_0038E',
       'DP05_0039E', 'DP05_0044E', 'DP05_0071E'],
      dtype='object')

In [72]:
us_states = [
    "Alabama", "Alaska", "Arizona", "Arkansas", "California",
    "Colorado", "Connecticut", "Delaware", "Florida", "Georgia",
    "Hawaii", "Idaho", "Illinois", "Indiana", "Iowa",
    "Kansas", "Kentucky", "Louisiana", "Maine", "Maryland",
    "Massachusetts", "Michigan", "Minnesota", "Mississippi", "Missouri",
    "Montana", "Nebraska", "Nevada", "New Hampshire", "New Jersey",
    "New Mexico", "New York", "North Carolina", "North Dakota", "Ohio",
    "Oklahoma", "Oregon", "Pennsylvania", "Rhode Island", "South Carolina",
    "South Dakota", "Tennessee", "Texas", "Utah", "Vermont",
    "Virginia", "Washington", "West Virginia", "Wisconsin", "Wyoming",
    "District of Columbia"
]
feature_df=pd.DataFrame()
for state in us_states:
    row1=df2017[df2017['NAME']==state]
    row2=df2018[df2018['NAME']==state]
    row3=df2019[df2019['NAME']==state]
    row4=df2021[df2021['NAME']==state]
    row5=df2022[df2022['NAME']==state]
    df=pd.concat([row1,row2,row3,row4,row5])
    df=df.drop('NAME',axis=1)
    df=df.set_index('Year').pct_change().reset_index()
    df['RegionName']=state
    feature_df=pd.concat([feature_df,df])
feature_df=feature_df.dropna()
feature_df['Year']=feature_df['Year'].astype(str)

In [73]:
feature_df.head()

Year  DP02_0001E  DP02_0002E  DP02_0003E  DP02_0007E  DP02_0011E  \
1  2018    0.007341    0.009226   -0.008186    0.090413    0.007228   
2  2019    0.022851   -0.265678   -0.360479   -0.317095   -0.802204   
3  2021    0.036880    0.013836    0.056032   -0.010988    0.081890   
4  2022    0.024848    0.041137    0.053954   -0.058821    0.009307   
1  2018    0.015195    0.000753   -0.043082   -0.078405    0.043289   

   DP02_0037E  DP02_0060E  DP02_0061E  DP02_0062E  ...  DP04_0134E  \
1    0.004783    0.022710   -0.002918    0.018174  ...    0.050667   
2    1.302990   -0.588360   -0.705200    0.439418  ...    0.024112   
3   -0.050374   -0.025077   -0.044056    0.038532  ...    0.066914   
4    0.106362   -0.070979   -0.062720   -0.020504  ...    0.060395   
1    0.942898   -0.208181    0.031255    0.017837  ...   -0.019983   

   DP05_0001E  DP05_0004E  DP05_0018E  DP05_0037E  DP05_0038E  DP05_0039E  \
1    0.002692    0.003198    0.010283   -0.001775   -0.000327   -0.123824   
2    0.003133   -0.006376    0.002545    0.005908    0.009572    0.054480   
3    0.027878    0.008556    0.010152   -0.013376   -0.010947    0.067268   
4    0.006829   -0.001060   -0.005025    0.006291   -0.002353   -0.017318   
1   -0.003186   -0.009174    0.011594   -0.000634    0.139775    0.012942   

   DP05_0044E  DP05_0071E  RegionName  
1   -0.027097    0.047111     Alabama  
2    0.015884    0.036934     Alabama  
3    0.040587    0.082710     Alabama  
4    0.146484   -0.125037     Alabama  
1   -0.060828    0.026241      Alaska  

[5 rows x 33 columns]

In [74]:
zillow=pd.read_csv('data/Zillow.csv')
zillowlong=zillow.melt(id_vars=['RegionName'],var_name='Year',value_name='Typical Home Value')

In [75]:
zillowlong.head()

RegionName  Year  Typical Home Value
0    California  2017       498316.221451
1         Texas  2017       192963.504413
2       Florida  2017       221751.216789
3      New York  2017       316846.572473
4  Pennsylvania  2017       174153.095534

In [76]:
target_df=pd.DataFrame()
for state in us_states:
    row1=zillowlong[(zillowlong['RegionName']==state)&(zillowlong['Year']=='2017')]
    row2=zillowlong[(zillowlong['RegionName']==state)&(zillowlong['Year']=='2018')]
    row3=zillowlong[(zillowlong['RegionName']==state)&(zillowlong['Year']=='2019')]
    row4=zillowlong[(zillowlong['RegionName']==state)&(zillowlong['Year']=='2021')]
    row5=zillowlong[(zillowlong['RegionName']==state)&(zillowlong['Year']=='2022')]
    df=pd.concat([row1,row2,row3,row4,row5])
    df=df.drop('RegionName',axis=1)
    df=df.set_index('Year').pct_change().reset_index()
    df['RegionName']=state
    target_df=pd.concat([target_df,df])

In [77]:
target_df=target_df.dropna()
final_data=pd.merge(feature_df,target_df,on=['RegionName','Year'])
new_column_names=['Year','Total households','Married couple households','Married couple with children of the householder under 18 years','Male householder,no spouse/partner present with children of the householder under 18 years','Female householder, no spouse/partner present with children of the householder under 18 years','Number of women 15 to 50 years old who had a birth in the past 12 months','Less than 9th grade','9th to 12th grade,no diploma','High school graduate (includes equivalency)','Some college, no degree','Associates degree','Bachelors degree','Number of people employed','Number of people unemployed','Median Household Income','Mean Household Income','Median earnings for male full-time, year-round workers','Median earnings for female full-time, year-round workers','Occupied Housing Units','Total Housing Units','Median Rooms','Renter Occupied Housing Units','Mean Rent Paid','Total population','Sex ratio (males per 100 females)','Median Age','Race-White','Race-Black','Race-American Indian and Alaska Native','Race-Asian','Hispanic or Latino','RegionName','Typical Home Value']
final_data.columns=new_column_names
final_data.to_csv('./data/RecordData.csv', index=False)

In [78]:
final_data.head()

Year  Total households  Married couple households  \
0  2018          0.007341                   0.009226   
1  2019          0.022851                  -0.265678   
2  2021          0.036880                   0.013836   
3  2022          0.024848                   0.041137   
4  2018          0.015195                   0.000753   

   Married couple with children of the householder under 18 years  \
0                                          -0.008186                
1                                          -0.360479                
2                                           0.056032                
3                                           0.053954                
4                                          -0.043082                

   Male householder,no spouse/partner present with children of the householder under 18 years  \
0                                           0.090413                                            
1                                          -0.317095                                            
2                                          -0.010988                                            
3                                          -0.058821                                            
4                                          -0.078405                                            

   Female householder, no spouse/partner present with children of the householder under 18 years  \
0                                           0.007228                                               
1                                          -0.802204                                               
2                                           0.081890                                               
3                                           0.009307                                               
4                                           0.043289                                               

   Number of women 15 to 50 years old who had a birth in the past 12 months  \
0                                           0.004783                          
1                                           1.302990                          
2                                          -0.050374                          
3                                           0.106362                          
4                                           0.942898                          

   Less than 9th grade  9th to 12th grade,no diploma  \
0             0.022710                     -0.002918   
1            -0.588360                     -0.705200   
2            -0.025077                     -0.044056   
3            -0.070979                     -0.062720   
4            -0.208181                      0.031255   

   High school graduate (includes equivalency)  ...  Total population  \
0                                     0.018174  ...          0.002692   
1                                     0.439418  ...          0.003133   
2                                     0.038532  ...          0.027878   
3                                    -0.020504  ...          0.006829   
4                                     0.017837  ...         -0.003186   

   Sex ratio (males per 100 females)  Median Age  Race-White  Race-Black  \
0                           0.003198    0.010283   -0.001775   -0.000327   
1                          -0.006376    0.002545    0.005908    0.009572   
2                           0.008556    0.010152   -0.013376   -0.010947   
3                          -0.001060   -0.005025    0.006291   -0.002353   
4                          -0.009174    0.011594   -0.000634    0.139775   

   Race-American Indian and Alaska Native  Race-Asian  Hispanic or Latino  \
0                               -0.123824   -0.027097            0.047111   
1                                0.054480    0.015884            0.036934   
2                                0.067268    0.040587            0.082710   
3                               -0.017318    0.1464

In [79]:
final_data.shape

(204, 34)

# Record Data Overview

As you can see with the cleaned record data above, my data set has 204 rows and 34 columns. Two of the columns serve as an index, with the state and year as the unique identifiers. Each cell represents the percent change of the variable from the previous year. For example, the first row is Alabama-2018. This row represents the percent change from 2017 to 2018 for all the census predictor variables and the home value target variable. More data would be ideal for my analysis, but for the purposes of this project I will continue with the data I already wrangled. 31 features are a lot and I should definitely aim to cut this number down. More on this in the dimensionality reduction and data exploration tabs, but I start below.

Now I will quickly look at the max of each column just to see if there are any obvious problems.

In [90]:
DP02colss=['DP02_0001E','DP02_0002E','DP02_0003E','DP02_0007E','DP02_0011E','DP02_0037E','DP02_0060E','DP02_0061E','DP02_0062E','DP02_0063E','DP02_0064E','DP02_0065E']
DP03colss=['DP03_0026PE','DP03_0005E','DP03_0062E','DP03_0063E','DP03_0093E','DP03_0094E']
DP04colss=['DP04_0002E','DP04_0001E','DP04_0037E','DP04_0047E','DP04_0134E']
DP05colss=['DP05_0001E','DP05_0004E','DP05_0018E','DP05_0037E','DP05_0038E','DP05_0039E','DP05_0044E','DP05_0071E']

In [91]:
for name in DP02colss:
    print(feature_df[name].max())

0.0960146791508043
0.08325965443347161
0.21027912686129047
0.5718313168095854
0.31098100619511193
4.71179314694135
0.5836723372781065
0.2331338411316648
1.16632039564257
3.7135790077430455
0.20096335286522993
1.8600379486638907


In [92]:
for name in DP03colss:
    print(feature_df[name].max())

0.08159905963183367
0.9855738883335392
0.1558987776230465
0.13903133903133913
0.18793842369265668
0.19923024054982807


In [93]:
for name in DP04colss:
    print(feature_df[name].max())

0.0960146791508043
0.10739311182290723
0.03921568627450989
0.1304236996210817
0.17613636363636354


In [94]:
for name in DP05colss:
    print(feature_df[name].max())

0.06371228802533757
0.024340770791075217
0.026666666666666616
0.04291548345376328
0.9405508590128171
1.598558282208589
0.3942459396751741
0.7167422884813359


Looks like there are some obvious outliers, with one cell having a 470% increase in the metric. With the ACS data, for each measure they release to the public, they also release a margin of error. The margin of error can be high at times, explaining why some percent change values may be out of the ordinary. In the next tab I will dig deeper into the outliers.